pkg/receive/handler.go

```diff
@@ -360,6 +360,8 @@ func (h *Handler) receiveHTTP(w http.ResponseWriter, r *http.Request) {
 		http.Error(w, err.Error(), http.StatusServiceUnavailable)
 	case errConflict:
 		http.Error(w, err.Error(), http.StatusConflict)
+	case errOutOfBounds:
+		http.Error(w, err.Error(), http.StatusBadRequest)
```

Comment:
Maybe I'm missing something here, but my understanding was that Prometheus treats as _recoverable_ only network errors, `5xx` responses and, if configured to, `429` responses (see [this line](https://github.com/prometheus/prometheus/blob/579331446a04503b2b1515767e7ceaceab2c6342/storage/remote/client.go#L213) and below).

It therefore should not make difference if we're responding with bad request (`400`) or  conflict (`409`), since conflict is already unrecoverable, right?

Reply:
For Prometheus, 400 = unrecoverable. If that's your question? Also, if that's untrue, than I'm happy to get pointed to some resource. However, I've tested this and the response now becomes unrecoverable for Prometheus.

Reply:
As far as I understood, _only_ `5xx` codes are recoverable for Prometheus (see the link in my comment). Meaning if until now we were returning `409` for out of bounds error, it should have already been working as expected, i.e. there should not be a real difference between the codes. As long as we return any `4xx` code, it should be treated as unrecoverable (with the possible exception of `429`).

Reply:
Right, I understand you and the code (missed the link) now. Funny thing is though, that I did _not_ experience this behaviour. Only after my change I actually see the non-recoverable errors in the Prometheus logs. 🤔

Reply:
More on this: https://github.com/prometheus/prometheus/pull/5649#issuecomment-970730016

Reply:
looking forward to seeing this fixed. it's causing us pain in production.

Reply:
@wiardvanrij I did a small test today: I spun up a VM where I skewed the clock on purpose (made it run in the 'past' compared to my host machine), I brought up an instance of Prometheus, configured to remote write to Thanos instance running on my host. 

I first run it against Thanos build from latest `main`. The result - Prometheus informs me this is a non-recoverable error:
```
ts=2022-02-12T12:26:39.089Z caller=dedupe.go:112 component=remote level=error remote_name=6c3063 url=http://192.168.56.1:19291/api/v1/receive msg="non-recoverable error" count=9 exemplarCount=0 err="server returned HTTP status 409 Conflict: store locally for endpoint 127.0.0.1:10901: conflict"
```
Built from your branch:
```
ts=2022-02-12T12:34:32.266Z caller=dedupe.go:112 component=remote level=error remote_name=6c3063 url=http://192.168.56.1:19291/api/v1/receive msg="non-recoverable error" count=63 exemplarCount=0 err="server returned HTTP status 400 Bad Request: store locally for endpoint 127.0.0.1:10901: out of bounds"
```

I haven't noticed any unexpected behavior on either side (Thanos or Prometheus). Prometheus treated it (in both cases) simply as unrecoverable error.

Reply:
thanks @matej-g  really appreciate the testing - Then I need more user input. @Kampe did you have any chance to test it or could you provide more info? Because if we now look at the code, it already should be correct. I would agree that we sometimes see odd behaviour but then we really need to know why that is happening. 

If I have some more time, i'll try to figure things out as well.

Reply:
I responded in the issue ticket https://github.com/thanos-io/thanos/issues/4831, but there definitely seems to be two issues happening with 500 level errors and 409's.  Unsure if they are related, but it seems we can on atleast a monthly cadence (sometimes weekly) can produce them.

We have not had the chance to roll the changes to any of our environments, we don't have much in the way of tooling setup to be able to roll non registry hosted third party tools to even developer environments so testing without a prebuilt container for us is pretty intensive, however when we have a container we can reference I can get taht rolled in a matter of minutes.